### PR TITLE
Use SearchValues throughout dotnet/aspnetcore (part 2)

### PR DESCRIPTION
**Pull request by ladeak**
# Use SearchValues throughout dotnet/aspnetcore

Using SearchValues in cookie header parser to find the end of the cookie value.

## Description

- Replacing the loop to search for invalid vales in CookieHeaderParserShared with SearchValues (one of the tasks in https://github.com/dotnet/aspnetcore/issues/46484 )
- Covering tests for all allowed values.

For the performance comparison, I used the existing `RequestCookieCollectionBenchmarks`:

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22631
12th Gen Intel Core i7-1255U, 1 CPU, 12 logical and 10 physical cores
.NET SDK=9.0.100-preview.3.24161.2
  [Host]     : .NET 9.0.0 (9.0.24.17301), X64 RyuJIT
  Job-XWCXZR : .NET 9.0.0 (9.0.24.16003), X64 RyuJIT

Server=True  Toolchain=.NET Core 9.0  InvocationCount=1
RunStrategy=Throughput  UnrollFactor=1
```

Before:

|              Method |     Mean |    Error |   StdDev |     Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------- |---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| Parse_TypicalCookie | 13.11 us | 1.670 us | 4.845 us | 76,275.9 |     - |     - |     - |      1 KB |

After:

|              Method |     Mean |     Error |    StdDev |      Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------- |---------:|----------:|----------:|----------:|------:|------:|------:|----------:|
| Parse_TypicalCookie | 6.914 us | 0.2298 us | 0.6444 us | 144,628.1 |     - |     - |     - |     952 B |

If the approach is agreed, I would continue to extend this PR covering further tasks mentioned in the linked issue. At that point I will mark the PR as draft until those are done. Although I can do independent PRs as well, please indicate the preference.

### CODE REVIEW

**Review by stephentoub — COMMENTED** [review]
Nice

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 1]
```diff
@@ -192,15 +198,14 @@ internal static StringSegment GetCookieValue(StringSegment input, ref int offset
             offset++;
         }
 
-        while (offset < input.Length)
+        var delta = input.AsSpan(offset).IndexOfAnyExcept(CookieValueChar);
+        if (delta == -1)
```
Comment (gfoidl) [inline src/Http/Shared/CookieHeaderParserShared.cs:202]:
```suggestion
        if (delta < 0)
```
Produces a tad better machine code for the CPU which can then fuse the comparison and the jump.

Comment (ladeak) [reply src/Http/Shared/CookieHeaderParserShared.cs:202]: I recall such myself too, although I was not able to measure any difference with the mentioned benchmark. Regardless, I apply the change, thank you for pointing out.

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 2]
```diff
@@ -9,6 +10,11 @@ namespace Microsoft.Net.Http.Headers;
 
 internal static class CookieHeaderParserShared
 {
+    // cookie-octet      = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
+    //                     ; US-ASCII characters excluding CTLs, whitespace DQUOTE, comma, semicolon, and backslash
```
Comment (amcasey) [inline src/Http/Shared/CookieHeaderParserShared.cs:14]:
```suggestion
    //                     ; US-ASCII characters excluding CTLs, whitespace, DQUOTE, comma, semicolon, and backslash
```

Comment (amcasey) [reply src/Http/Shared/CookieHeaderParserShared.cs:14]: (I know it was just copy-paste :wink:)

**Review by amcasey — APPROVED** [review]

**Review by amcasey — APPROVED** [review]
Thanks for the update!  The CI failure is unrelated, so I've kicked off a re-run.

**Comment by amcasey** [comment]
Anything else before I merge, @gfoidl?

**Review by gfoidl — APPROVED** [review]

**Pull request merged**